### PR TITLE
Add phase2 Inner Tracker (IT) pixel cluster to cmssw 81X from new IB

### PR DESCRIPTION
**Pull request by emiglior**
@atricomi @ebrondol 
This PR superseeds the PR #14854 

We introduce a DataFormats + producer for the Inner Tracker pixel clusters for phase2.
The new data format is essentially a clone of the present DataFormats/SiPixelCluster.
The reasons for introducing the new data format are:
-  the current data format for the pixel cluster makes use of types, e.g. https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DataFormats/SiPixelCluster/interface/SiPixelCluster.h#L77-L87 which optimize the performance with the current detector but do not allow to span the full module in the "small pitch scenario" foreseen for phase2 pixels (e.g. pixel with column or row index larger than 2**8-1=255);
- more in general, the phase2 specs are still under definition and simulation studies are part of this process. So we would like to be able to change the clusterizer without interfering with the code in use for current/phase1 detector.

The new collection is included in the phase2 workflows but  for the moment no rechits are produced out of it.

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by emiglior** [comment]
@boudoul @delaere This is of interest for you as well

**Comment by civanch** [comment]
@emiglior , few minor comments:
in Phase2ITPixelCluster.h - there is no sense to keep '#ifndef CMS_NOCXX11' now, all CMS production compilers are c++11
in Phase2ITTPixelThresholdClusterizer.cc in line #228 use std::cout may be a problem, lines #361-372 may be removed
in test/Phase2ITPixelClusterValidation in line #67 better to inherit from edm::one::EDAnalyser

**Comment by emiglior** [comment]
@civanch  maybe mine are naive questions:

> in Phase2ITTPixelThresholdClusterizer.cc in line #228 use std::cout may be a problem

because of the cout (to be replaced with a MessageLogger fcn) or because of the if/then structure?

> in test/Phase2ITPixelClusterValidation in line #67 better to inherit from edm:1️⃣:EDAnalyser

maybe there is some issue with the fonts: do you really mean edm::1::EDAnalyser?

**Comment by kpedro88** [comment]
@emiglior:

couts are not allowed in normal operation of CMSSW. You can either use a preprocessor debug flag (which is turned off by default), or use MessageLogger.

I think it should be `edm::one::EDAnalyzer`.

**Comment by civanch** [comment]
@emiglior , @kpedro88 has responded correctly, my browser somehow substitute 'one'. Also std::cout cannot be used in a regular code anymore.

**Comment by emiglior** [comment]
apologies for the bad commit d822c44

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by emiglior** [comment]
thx for starting the comparison

**Comment by civanch** [comment]
+1

**Comment by emiglior** [comment]
Do I interpret correctly that the "technical" tests have been passed?
What are then the next steps?

**Comment by kpedro88** [comment]
@davidlange6 how do we handle "new-package-pending"?

**Comment by slava77** [comment]
@kpedro88  edit https://github.com/cms-sw/cms-bot/blob/master/categories.py to assign new code to corresponding category

@cvuosalo this replaces #14854 and now doesn't have the reco signature (yet)

**Comment by kpedro88** [comment]
@slava77 https://github.com/cms-sw/cms-bot/pull/715 is merged now, how do we tell the bot to update the labels?

**Comment by kpedro88** [comment]
@slava77 okay, apparently there's two new packages here.

**Comment by smuzaffar** [comment]
@kpedro88 , just adding a comment should wake up bot to re-assign labels :-)

**Comment by kpedro88** [comment]
@slava77 you can sign now

**Comment by slava77** [comment]
On 6/20/16 3:02 PM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> you can sign now

It's on Carl's list.

**Comment by kpedro88** [comment]
@cvuosalo ping

**Comment by kpedro88** [comment]
@davidlange6 we would like this in pre8...

**Comment by kpedro88** [comment]
@cvuosalo @davidlange6 ping

**Comment by davidlange6** [comment]
this one needs a reco review (and maybe a reco discussion to understand on the topic I might propose)

**Comment by kpedro88** [comment]
This is not a well-functioning system if a PR submitted 2 days after pre7 was released cannot get into pre8.

**Comment by davidlange6** [comment]
Hi @kpedro88 - I agree things can be better -but as to distinguish rules and exceptions -  I think a more appropriate complaint is that there are no new issues raised to address since 1 week in this PR. Maybe that is not so terrible for 2k lines of code. As usual changes involving new packages take extra time to get to the right set of people

**Comment by cvuosalo** [comment]
@kpedro88: Sorry for the delay. CMS Week at CERN has taken up a lot of time. I am reviewing this PR now.
For future PRs, if the developer includes in the PR description some priority or target release (e.g., "needs to go into 810pre8"), that would be helpful to prioritize review. It is good to document which PRs are essential for further development or are holding up other PRs.

File: DataFormats/Phase2ITPixelCluster/interface/Phase2ITPixelCluster.h [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+//!  Class to contain and store all the topological information of pixel clusters:
+//!  charge, global size, size and the barycenter in x and y
+//!  local directions. It builds a vector of SiPixel (which is
+//!  an inner class) and a container of channels. 
+//!
+//!  March 2007: Edge methods moved to RectangularPixelTopology class (V.Chiochia)
+//!  Feb 2008: Modify the Pixel class from float to shorts
+//!  May   2008: Offset based packing (D.Fehling / A. Rizzi)
+//!  Sep 2012: added Max back, removed detId (V.I.)
+//!  sizeX and sizeY now clipped at 127
```
Comment (cvuosalo) [inline DataFormats/Phase2ITPixelCluster/interface/Phase2ITPixelCluster.h]: These history comments are copied from the original file and are irrelevant and confusing here. Please delete them.

File: DataFormats/Phase2ITPixelCluster/interface/Phase2ITPixelCluster.h [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+      if (ysize > maxCol) maxCol = ysize;
+    }
+    // assume minimum is correct
+    uint32_t minCol= minPixelCol();
+    packCol(minCol,maxCol);
+    uint32_t minRow= minPixelRow();
+    packRow(minRow,maxRow);
+  }
+  
+  // ggiurgiu@fnal.gov, 01/05/12
```
Comment (cvuosalo) [inline DataFormats/Phase2ITPixelCluster/interface/Phase2ITPixelCluster.h]: All the false history comments in all the files should be deleted. In the future, they could confuse readers about the age and origin of these files.

**Comment by cvuosalo** [comment]
As was discussed previously (https://github.com/cms-sw/cmssw/pull/14854#issuecomment-225654512), the Reco conveners have concerns about massive copying of hundreds of lines of code and then applying simple search-and-replace to make minor customizations. C++ has very powerful inheritance mechanisms explicitly designed to eliminate the need for such copying. A developer would need to make a strong case to justify abandonment of inheritance in favor of brute-force copying, but I haven't seen a sufficient justification. It would be good to have a presentation at the Reco meeting Wed., June 29, to explain and justify the design choices made in this PR.
This issue need not hold up this PR, but it will need to be addressed soon, perhaps in a future PR.

**Comment by VinInn** [comment]
as simple as DO NOT EVEN THINK TO INTERFERE WITH ONGOING RUN2 AND PHASE1 ACTIVITIES

I do not what to see even a single added virtual or not inlined method in code currently used w/o a full assessment of its performance and impact.
Given that at the moment PHASE2 is in a R&D phase and most of this code may change easily even at next iteration the best engineering practice it to keep it separated. Once consolidated we will discuss how to merge it with mainstream code.

**Comment by cvuosalo** [comment]
@VinInn: I am not suggesting any changes at all to existing code. If there are unavoidable roadblocks to employing inheritance or if the base classes are subject to frequent changes for optimization of current performance, then those would be justifications for copying over inheritance. This type of question is exactly the kind of topic we would like discussed in Reco meetings. In addition, we would like to hear a plan and timeline for the development of this Phase 2 code and its eventual integration into the mainstream codebase.

**Comment by atricomi** [comment]
I fully sign what Vincenzo said.
The reason, as we already explained, to have a different data format (and all modification herein) is that for the small pitches we certainly cannot run with the present code. Either we modify it (but this means to touch the code which currently is used in the data taking and also to spoil its performance) or we proceed in parallel avoiding any interference and gaining also the flexibility to follow the needed change that comes from hardware suggestions. We still don’t have a final design for the phase2 pixel, we even don’t know if the analog readout would be maintained or not, so we need to have flexibility without causing any impact in the actual code.
Surely improvement in the design and performance will be possible in the future but we first need to have something able to digest different detector configuration. Once the detector design will be consolidated we may think also to the code optimization.

Il giorno 27/giu/2016, alle ore 11:15, Vincenzo Innocente <notifications@github.com> ha scritto:

File: DataFormats/Phase2ITPixelCluster/src/Phase2ITPixelCluster.cc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+//!  March 2007: Edge pixel methods moved to RectangularPixelTopology (V.Chiochia)
+//!  May   2008: Offset based packing (D.Fehling / A. Rizzi)  
+//!  \author Petar Maksimovic, JHU
+//---------------------------------------------------------------------------
+
+
+Phase2ITPixelCluster::Phase2ITPixelCluster( const Phase2ITPixelCluster::PixelPos& pix, uint32_t adc) :
+  thePixelRow(pix.row()),
+  thePixelCol(pix.col()),
+    // ggiurgiu@fnal.gov, 01/05/12
```
Comment (cvuosalo) [inline DataFormats/Phase2ITPixelCluster/src/Phase2ITPixelCluster.cc]: @emiglior: There are still several false-history comments that should be deleted in the various files in this PR. Should I flag them all?

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
Phase 2 workflow test in progress...

**Comment by cvuosalo** [comment]
@emiglior: Could you please provide a list of how the C++ header files in this PR are used? That is, where are they used or included? You could post the list here or show it in the Reco meeting on 29 June.
Thanks.

**Comment by cvuosalo** [comment]
+1

For #14885 030499c937369d987b7058ab4ada21ffe2aba72c

Adding Phase 2 inner tracker pixel clusters. There should be no change in standard workflows.

The new code is marginally acceptable as provisional code that will be re-worked as detector details are finalized, but it will require significant revision to merge it into the mainstream codebase. Jenkins tests against baseline CMSSW_8_1_X_2016-06-27-1100 show no significant differences, as expected. A test of workflow 11024.0_TTBar (2023) with 20 events against baseline CMSSW_8_1_X_2016-06-19-2300 also shows no differences in existing quantities. CPU time increases marginally for the new module (first event excluded from measurement):
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.24%         0.00 ms/ev ->         1.53 ms/ev phase2ITPixelClusters
   -0.274684      -0.06%         1.48 ms/ev ->         1.12 ms/ev siPixelClusters
  ---------- ------------     --------                  ----       ------------
Job total:  0.623225 s/ev ==> 0.662135 s/ev
```
RECO event size also increases as might be expected:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->     26202.8      26203     NEWO   7.87     Phase2ITPixelClusteredmNewDetSetVector_phase2ITPixelClusters__RECO.
-------------------------------------------------------------
   332934 ->      359129      26195             7.9     ALL BRANCHES
```

**Comment by civanch** [comment]
+1

**Comment by emiglior** [comment]
Hi, should we assume that the decision about this PR was positive?
Please let us know in order to make a realistic plan for phase2 deliverables in 81X.

**Comment by kpedro88** [comment]
@slava77 @cvuosalo @davidlange6 let's merge if possible...

**Comment by emiglior** [comment]
Great!

**Comment by kpedro88** [comment]
@davidlange6 are we going to merge this?

**Comment by VinInn** [comment]
Where are we with this PR?
we need it integrated to get the CPE in and start working to use these clusters in OmniCluster, PixelRecHit and reco code

**Comment by emiglior** [comment]
ping... 
We really need this PR in the next pre-release

**Comment by kpedro88** [comment]
@davidlange6 ...

**Comment by davidlange6** [comment]
+1
@vininn and I discussed some options forward, but will approve this for now so that next steps happen in parallel. I the meanwhile, undoubtedly this will trigger much more code copying downstream that hopefully can be avoided with some thought by the development team.

**Comment by emiglior** [comment]
Thx to all the involved!
Does it mean that will be included in  810_pre9?

**Comment by kpedro88** [comment]
@emiglior yes, pre9 has not been built yet, so all PRs merged so far will be included.

**Pull request merged**